File: scriptSpeed/lua/scripts/pointSimul.lua
---------------------------------------------
-- all computation for one point
--
function allScript(point,mouse,friction)
	-- some call saving precomputation
	diffX = mouse:getX() - point:getX();
	diffY = mouse:getY() - point:getY();

	-- modify velocity based on point decision
	distance = (diffX^2 + diffY^2)^0.5;
	if distance > 20 then
		point:setVelocityX(point:getVelocityX() + (diffX) / distance);
		point:setVelocityY(point:getVelocityY() + (diffY) / distance);
	elseif distance < 15 then
		point:setVelocityX(point:getVelocityX() - (diffX) / distance);
		point:setVelocityY(point:getVelocityY() - (diffY) / distance);
	end

	-- let the physics happen
	speed = (point:getVelocityX()^2 + point:getVelocityY()^2)^0.5;
	if speed <= friction then
		point:setVelocityX(0);
		point:setVelocityY(0);
	else
		-- apply friction
		point:setVelocityX((1 - friction / speed) * point:getVelocityX());
		point:setVelocityY((1 - friction / speed) * point:getVelocityY());

		-- move points
		point:setX(point:getX() + point:getVelocityX());
		point:setY(point:getY() + point:getVelocityY());
	end
end

---------------------------------------------
-- half of the computation for one point, rest delegated to native call
--
function scriptToNative(point,mouse,friction)
	-- some call saving precomputation
	diffX = mouse:getX() - point:getX();
	diffY = mouse:getY() - point:getY();

	-- modify velocity based on point decision
	distance = (diffX^2 + diffY^2)^0.5;
	if distance > 20 then
		point:setVelocityX(point:getVelocityX() + (diffX) / distance);
		point:setVelocityY(point:getVelocityY() + (diffY) / distance);
	elseif distance < 15 then
		point:setVelocityX(point:getVelocityX() - (diffX) / distance);
		point:setVelocityY(point:getVelocityY() - (diffY) / distance);
	end

	stepPoint(point);
end

---------------------------------------------
-- all computation for one point with minimal number of native calls
--
function allScriptMinCallback(point,mouse,friction)
	-- some call saving precomputation
	x = point:getX();
	y = point:getY();
	xVelocity = point:getVelocityX();
	yVelocity = point:getVelocityY();
	mouseX = mouse:getX();
	mouseY = mouse:getY();
	diffX = mouseX - x;
	diffY = mouseY - y;

	-- modify velocity based on point decision
	distance = (diffX^2 + diffY^2)^0.5;
	if distance > 20 then
		xVelocity = xVelocity + diffX / distance;
		yVelocity = yVelocity + diffY / distance;
	elseif distance < 15 then
		xVelocity = xVelocity - diffX / distance;
		yVelocity = yVelocity - diffY / distance;
	end

	-- let the physics happen
	speed = (xVelocity^2 + yVelocity^2)^0.5;
	if speed <= friction then
		point:setVelocityX(0);
		point:setVelocityY(0);
	else
		-- apply friction
		point:setVelocityX((1 - friction / speed) * xVelocity);
		point:setVelocityY((1 - friction / speed) * yVelocity);

		-- move points
		point:setX(x + xVelocity);
		point:setY(y + yVelocity);
	end
end

---------------------------------------------
-- lua script controls the simulation loop
--
function loopControl(pointCount, cyclesCount, friction)
	-- initialize the simulation
	initSimul();

	-- start main loop
	for i=0,cyclesCount-1,1 do

		-- process input and if the app should quit, break
		if processInput() then
			break;
		end

		for j=0,pointCount-1,1 do
			allScript(getPoint(j), getMouse(), friction);
		end

		draw();

	end

	-- clean resources
	cleanup();
end


---------------------------------------------
-- lua script controls the simulation loop and caches all point objects
--
function loopPointsCached(pointCount, cyclesCount, friction)
	-- initialize the simulation
	initSimul();

	points = {}
    for i=0,pointCount-1 do
      points[i] = getPoint(i);
    end

	-- start main loop
	for i=0,cyclesCount-1,1 do

		-- process input and if the app should quit, break
		if processInput() then
			break;
		end

		mouse = getMouse();

		for j=0,pointCount-1,1 do
			allScript(points[j], mouse, friction);
		end

		draw();

	end

	-- clean resources
	cleanup();
end

File: scriptSpeed/js/scripts/pointSimul.js
function allScript(point, mouse, friction) {
    // some call saving precomputation
    var diffX = mouse.x - point.x;
    var diffY = mouse.y - point.y;

    // modify velocity based on point decision
    var distance = Math.pow(Math.pow(diffX, 2) + Math.pow(diffY, 2), 0.5);
    if (distance > 20) {
        point.velocityX += diffX / distance;
        point.velocityY += diffY / distance;
    } else if (distance < 15) {
        point.velocityX -= diffX / distance;
        point.velocityY -= diffY / distance;
    }

    // let the physics happen
    var speed = Math.pow(Math.pow(point.velocityX, 2) + Math.pow(point.velocityY, 2), 0.5);
    if (speed <= friction) {
        point.velocityX = 0;
        point.velocityY = 0;
    } else {
        // apply friction
        point.velocityX = ((1 - friction / speed) * point.velocityX);
        point.velocityY = ((1 - friction / speed) * point.velocityY);

        // move points
        point.x += point.velocityX;
        point.y += point.velocityY;
    }
}

function scriptToNative(point, mouse, friction) {
    // some call saving precomputation
    var diffX = mouse.x - point.x;
    var diffY = mouse.y - point.y;

    // modify velocity based on point decision
    var distance = Math.pow(Math.pow(diffX, 2) + Math.pow(diffY, 2), 0.5);
    if (distance > 20) {
        point.velocityX += diffX / distance;
        point.velocityY += diffY / distance;
    } else if (distance < 15) {
        point.velocityX -= diffX / distance;
        point.velocityY -= diffY / distance;
    }

    stepPoint(point);
}

function allScriptMinCallback(point, mouse, friction) {
    // some call saving precomputation
    var x = point.x;
    var y = point.y;
    var velocityX = point.velocityX;
    var velocityY = point.velocityY;
    var diffX = mouse.x - x;
    var diffY = mouse.y - y;

    // modify velocity based on point decision
    var distance = Math.pow(Math.pow(diffX, 2) + Math.pow(diffY, 2), 0.5);
    if (distance > 20) {
        velocityX += diffX / distance;
        velocityY += diffY / distance;
    } else if (distance < 15) {
        velocityX -= diffX / distance;
        velocityY -= diffY / distance;
    }

    // let the physics happen
    var speed = Math.pow(Math.pow(velocityX, 2) + Math.pow(velocityY, 2), 0.5);
    if (speed <= friction) {
        point.velocityX = 0;
        point.velocityY = 0;
    } else {
        // apply friction
        point.velocityX = ((1 - friction / speed) * velocityX);
        point.velocityY = ((1 - friction / speed) * velocityY);

        // move points
        point.x = x + velocityX;
        point.y = y + velocityY;
    }
}

